server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -463,7 +466,12 @@ public TermsEnum getTerms(boolean caseInsensitive, String string, SearchExecutio
             throws IOException {
             IndexReader reader = queryShardContext.searcher().getTopReaderContext().reader();
 
-            Terms terms = MultiTerms.getTerms(reader, name());
+            Terms terms = null;
+            if (isIndexed()) {
+                terms = MultiTerms.getTerms(reader, name());
+            } else if (hasDocValues()) {
+                terms = SortedSetDocValuesTerms.getTerms(reader, name());
```

Comment:
Is there a reason why `MultiDocValues.getSortedSetValues(reader, name()).termsEnum()` wouldn't work?

Reply:
First of all, thanks for pointing me to that class (I wasn't aware of it).
- `MultiDocValues.getSortedSetValues` looks to build global ordinals (which `MultiTerms` doesn't do), i.e. more effort up-front.
- `MultiTerms` exposes the terms enum via `MultiTermsEnum`, which does a merge sort on retrieval by term text.
What would you recommend?

Reply:
Ah, I was thinking "wait we already have logic to merge multiple doc values instances", I had forgotten that it would build global ordinals under the hood. Merging terms on the fly rather than up-front makes more sense indeed.

Reply:
Can you leave a comment about why it's not using MultiDocValues?